From 1a3dd87e8d056109e19ef39ddc387ea42e399a49 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Wed, 8 Apr 2015 15:40:54 +0800
Subject: [PATCH] refpolicy: allow login_pgm manage user keys

Reference sources: selinux-policy-3.12.1-196.fc20.src.rpm
Changes come from: policy-f20-base.patch

* allow login_pgm manage keys for all user domains

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/system/authlogin.if | 20 ++++++++++++++++++++
 policy/modules/system/authlogin.te |  4 ++++
 2 files changed, 24 insertions(+)

diff --git a/policy/modules/system/authlogin.if b/policy/modules/system/authlogin.if
index e8e27c0..38fd6f9 100644
--- a/policy/modules/system/authlogin.if
+++ b/policy/modules/system/authlogin.if
@@ -105,9 +105,11 @@ interface(`auth_use_pam',`
 interface(`auth_login_pgm_domain',`
 	gen_require(`
 		type var_auth_t, auth_cache_t;
+		attribute login_pgm;
 	')
 
 	domain_type($1)
+	typeattribute $1 login_pgm;
 	domain_subj_id_change_exemption($1)
 	domain_role_change_exemption($1)
 	domain_obj_id_change_exemption($1)
@@ -1903,6 +1905,24 @@ interface(`auth_unconfined',`
 	typeattribute $1 can_relabelto_shadow_passwords;
 ')
 
+########################################
+## <summary>
+##     Manage keys for all user domains.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`userdom_manage_all_users_keys',`
+	gen_require(`
+		attribute userdomain;
+	')
+
+	allow $1 userdomain:key manage_key_perms;
+')
+
 #######################################
 ## <summary>
 ##      Manage create logins log.
diff --git a/policy/modules/system/authlogin.te b/policy/modules/system/authlogin.te
index f80dfcb..873072a 100644
--- a/policy/modules/system/authlogin.te
+++ b/policy/modules/system/authlogin.te
@@ -17,6 +17,7 @@ attribute can_read_shadow_passwords;
 attribute can_write_shadow_passwords;
 attribute can_relabelto_shadow_passwords;
 attribute nsswitch_domain;
+attribute login_pgm;
 
 type auth_cache_t;
 logging_log_file(auth_cache_t)
@@ -464,3 +465,6 @@ optional_policy(`
 	samba_read_var_files(nsswitch_domain)
 	samba_dontaudit_write_var_files(nsswitch_domain)
 ')
+
+# manage keys for all user domains
+userdom_manage_all_users_keys(login_pgm)
-- 
1.9.1

